rumqttd/src/protocol/mod.rs

```diff
@@ -616,32 +632,39 @@ pub fn valid_topic(topic: &str) -> bool {
 /// Checks if the filter is valid
 ///
 /// <https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718106>
-pub fn valid_filter(filter: &str) -> bool {
-    if filter.is_empty() {
+pub fn valid_filter(filter: impl AsRef<str>) -> bool {
+    let filter = filter.as_ref();
+    if !is_valid_topic_or_filter(filter) {
         return false;
     }
 
-    let hirerarchy = filter.split('/').collect::<Vec<&str>>();
-    if let Some((last, remaining)) = hirerarchy.split_last() {
-        for entry in remaining.iter() {
-            // # is not allowed in filter except as a last entry
-            // invalid: sport/tennis#/player
-            // invalid: sport/tennis/#/ranking
-            if entry.contains('#') {
-                return false;
-            }
-
-            // + must occupy an entire level of the filter
-            // invalid: sport+
-            if entry.len() > 1 && entry.contains('+') {
-                return false;
-            }
+    // rev() is used so we can easily get the last entry
+    let mut hirerarchy = filter.split('/').rev();
+
+    // split will never return an empty iterator
+    // even if the pattern isn't matched, the original string will be there
+    // so it is safe to just unwrap here!
+    let last = hirerarchy.next().unwrap();
```

Comment:
The `unwrap` here can easily be avoided with a `let else` that returns `false`. yes - the unwrap is safe but maybe you want to forbid unwrap in the whole crate at some point in time and avoiding the use of `unwrap` make the pile of work less.